src/module.c

```diff
@@ -13824,11 +13824,31 @@ int RM_RegisterDefragCallbacks(RedisModuleCtx *ctx, RedisModuleDefragFunc start,
  * When stopped and more work is left to be done, the callback should
  * return 1. Otherwise, it should return 0.
  *
- * NOTE: Modules should consider the frequency in which this function is called,
- * so it generally makes sense to do small batches of work in between calls.
+ * NOTE: We use certain thresholds to avoid excessive system calls.
+ * Time checks are only performed when any threshold is reached,
+ * which means we might slightly exceed the expected end time.
  */
 int RM_DefragShouldStop(RedisModuleDefragCtx *ctx) {
-    return (ctx->endtime != 0 && ctx->endtime <= getMonotonicUs());
+    if (ctx->stopping) /* Return immediately if already stopping */
+        return 1;
+    if (!ctx->endtime) /* Return if no time limit set */
+        return 0;
+        
+    /* Check time when any threshold is reached. */
+    if (++ctx->ops_count > 128 ||
```

Comment:
I think it's still necessary, and if the full defragmentation ends and there's no more new traffic, we may never quit.